File: Regular16/3.rb
#!/usr/bin/ruby -Ku

x = gets().chomp.split(/\s/)
idol_num = x[0].to_i
slot_num = x[1].to_i
costs = Array.new(slot_num, 0)
idols = Array.new()
score = Array.new(1<<idol_num, 10000000)
score[0] = 0

for i in 1..slot_num
  x = gets().chomp.split(/\s/)
  idol = Array.new(x[0].to_i)
  costs[i] = x[1].to_i
  for j in 1..x[0].to_i
    tmp = gets().chomp.split(/\s/)
    idol[j-1] = [tmp[0].to_i - 1, tmp[1].to_i]
  end
  idols << idol
end

for i in 1..1<<idol_num
  idols.each_with_index do |slot, snum|
    tmp = 1 * costs[snum+1]
    pcount = 0.0;

    slot.each do |tmpidol|
      if (i & (1<<tmpidol[0])) != 0
        tmp = tmp + score[i ^ (1<<tmpidol[0])] * tmpidol[1] / 100.0
        pcount += tmpidol[1].to_f
      end
    end

    next if pcount == 0.0
    tmp = tmp * 100 / pcount
    if score[i] > tmp
      score[i] = tmp
    end
  end
end

puts score[(1<<idol_num) - 1]

File: Regular16/3.c
#include <stdio.h>

double dp[1024];
int c[5],cost[5],idol[5][12],p[5][12];

int main(){
    int n,m,i,j,k,cnt;
    double tmp;
    scanf("%d%d",&n,&m);
    for(i=0;i<m;i++){
        scanf("%d%d",&c[i],&cost[i]);
        for(j=0;j<c[i];j++){
            scanf("%d%d",&idol[i][j],&p[i][j]);
            idol[i][j]--;
        }
    }
    dp[0]=0.0;
    for(i=1;i<(1<<n);i++){
        dp[i]=100000000000.0;
        for(j=0;j<m;j++){
            tmp=1.0*cost[j];cnt=0;
            for(k=0;k<c[j];k++){
                if(i&(1<<idol[j][k])){
                    tmp=tmp+dp[i^(1<<idol[j][k])]*p[j][k]/100.0;
                    cnt+=p[j][k];
                    printf("%d, %d, %d\n", i, idol[j][k], cnt);
                    printf("dp: %f\n", dp[i^(1<<idol[j][k])]);
                }
            }
            if(!cnt)continue;
            tmp=tmp*100.0/cnt;
            if(dp[i]>tmp)dp[i]=tmp;
        }
    }
    for(i=1;i<(1<<n);i++)
        printf("%f\n", dp[i]);
    printf("%.10lf\n",dp[(1<<n)-1]);
    return 0;
}
